README.md
# A Browser Cookie

[![Cookie Action Status](https://github.com/dbpkgs/cookie/workflows/npm-publish/badge.svg)](https://github.com/dbpkgs/cookie/actions) [![Downloads Action Status](https://img.shields.io/npm/dt/@dbpkgs/cookie.svg)](https://www.npmjs.com/package/@dbpkgs/cookie)

A lightweight browser cookie library for frontend applications <br/>
Library also supports special characters encoding

## Installation

```bash
npm install @dbpkgs/cookie
```

or

```
yarn add @dbpkgs/cookie
```

## Usage

### Set Cookie

```ts
import cookie from '@dbpkgs/cookie';

cookie.set('COOKIE_NAME', 'COOKIE_VALUE', options);
//options are optional, check below for acceptable options
```

#### Cookie options

```ts
{
  expires?: Date
  path?: string
  domain?: string
  secure?: boolean
}
```

### Get Cookie

```ts
import cookie from '@dbpkgs/cookie';

cookie.get('COOKIE_NAME');
```

### Remove Cookie

```ts
import cookie from '@dbpkgs/cookie';

cookie.remove('COOKIE_NAME');
```

## Example Usage

### Testing in applications

```ts
import cookie from '@dbpkgs/cookie';

cookie.set('session', 'Uxc70_67gGuHHvAmTy10a', {
  expires: new Date(2022, 03, 13),
  path: '',
  secure: true,
});
```

### Testing in unit test libraries

At the moment the library is being tested with jest unit testing library

```ts
import cookie from '@dbpkgs/cookie';

test('Check if cookie is set to browser properly', () => {
  expect(cookie.set('test_session', 'Ab7MNgGyql89hpPalIdgql01gTjkaGb5')).toBe(
    'test_session=Ab7MNgGyql89hpPalIdgql01gTjkaGb5',
  );
});
```

## License

MIT


CHANGELOG.md
# CHANGELOG

## 1.2.3 (May 28, 2023)

- Add support for node engines 14.17.1 and higher
- Upgrade dev dependencies

## 1.2.2 (Mar 26, 2023)

- Tighten validation of parameters by adding actual field validation rather than depending on typescript types

## 1.2.1 (May 10, 2022)

- Add support for applications with both server and client rendering like nextjs and gatsby
- Bug fix for multiple cookie values

## 1.1.0 (Feb 20, 2022)

- Tighten unit tests for testing libraries
- Add support for special characters encoding e.g `/ ? : @ & = + $`

## 1.0.9 (Feb 19, 2022)

- Fix bugs due to broken modules and types during imports
- Initial public release


src/__tests__/cookie.test.ts
import cookie from '../index';
import Cookie, { encode, decode, errorsSchema } from '../Cookie';

describe('"encode"', () => {
  it('should be a properly defined encode function', () => {
    expect(encode).toBeDefined();
    expect(typeof encode).toBe('function');
  });

  it('should return a proper encoded string', () => {
    expect(encode('/ ? : @ & = + $')).toBe('%2F%20%3F%20%3A%20%40%20%26%20%3D%20%2B%20%24');
  });
});

describe('"decode"', () => {
  it('should be a properly defined decode function', () => {
    expect(decode).toBeDefined();
    expect(typeof decode).toBe('function');
  });

  it('should return correct decoded value from an encoded string', () => {
    const encodedValue = encode('https://example.com');
    const decodedValue = decode(encodedValue);
    expect(decodedValue).toBe('https://example.com');
  });
});

describe('"Cookie"', () => {
  it('should be properly defined', () => {
    const cookieNull = new Cookie();

    expect(Cookie).toBeDefined();
    expect(cookieNull).toBeInstanceOf(Cookie);
  });

  it('should return proper cookie results even if no document is available', () => {
    const cookieNull = new Cookie();

    expect(cookieNull.set('Hello', 'Hello')).toBe('Hello=Hello');
    expect(cookieNull.get('Hello')).toBe('Hello');
  });

  it('should return proper results even if no valid document object was provided', () => {
    const cookieNull = new Cookie('Hello');

    expect(cookieNull.set('Hello', 'Hello')).toBe('Hello=Hello');
    expect(cookieNull.get('Hello')).toBe('Hello');
  });
});

describe('"cookie.get"', () => {
  it('should be a properly defined cookie.get function', () => {
    expect(cookie.get).toBeDefined();
    expect(typeof cookie.get).toBe('function');
  });

  it('should return null when getting a cookie that does not exists in cookie storage', () => {
    expect(cookie.get('MY_COOKIE')).toBeNull();
  });

  it('should return a valid cookie when getting a cookie already set in cookie storage', () => {
    expect(cookie.get).toBeDefined();
    expect(cookie.set).toBeDefined();
    expect(cookie.set('test_session', 'Ab7MNgGyql89hpPalIdgql01gTjkaGb5')).toBe(
      'test_session=Ab7MNgGyql89hpPalIdgql01gTjkaGb5',
    );
    expect(cookie.get('test_session')).toBe('Ab7MNgGyql89hpPalIdgql01gTjkaGb5');
  });

  it('should set a cookie to cookie storage and return null when getting a cookie with the wrong key', () => {
    const sessionCookie = cookie.set('test_session', 'Ab7MNgGyql89hpPalIdgql01gTjkaGb5');

    expect(cookie.set).toBeDefined();
    expect(sessionCookie).toBe('test_session=Ab7MNgGyql89hpPalIdgql01gTjkaGb5');
    expect(cookie.get('test-sessions')).toBeNull();
  });

  it('should get the decoded cookie and return a proper cookie string when a cookie was set with options', () => {
    const sessionCookie = cookie.set('test_session', 'Ab7MNgGyql89hpPalIdgql01gTjkaGb5', {
      expires: new Date('Fri Apr 15 2022 '),
      path: '/',
      secure: true,
      domain: 'https://example.com',
    });

    expect(cookie.set).toBeDefined();
    expect(sessionCookie).toBe(
      `test_session=Ab7MNgGyql89hpPalIdgql01gTjkaGb5; expires=${new Date(
        'Fri Apr 15 2022',
      )}; path=%2F; domain=https%3A%2F%2Fexample.com; secure`,
    );

    expect(cookie.get).toBeDefined();
    expect(cookie.get('test_session')).toBe('Ab7MNgGyql89hpPalIdgql01gTjkaGb5');
  });

  it('should throw an error when getting a cookie value with a non-string key', () => {
    expect(() => {
      // Explicitly force use of non-string key when getting a token
      //@ts-expect-error (2345) :argument of type 'number' is not assignable to parameter of type 'string'.
      cookie.get(1);
    }).toThrowError(errorsSchema.key);
  });
});

describe('"cookie.set"', () => {
  it('should be a properly defined cookie.set function', () => {
    expect(cookie.set).toBeDefined();
    expect(typeof cookie.set).toBe('function');
  });

  it('should set the cookie properly in the cookie storage and return the set cookie', () => {
    const sessionCookie = cookie.set('test_session', 'Ab7MNgGyql89hpPalIdgql01gTjkaGb5');

    expect(sessionCookie).toBe('test_session=Ab7MNgGyql89hpPalIdgql01gTjkaGb5');
  });

  it('should return a proper cookie string when getting a cookie set cookie with cookie options', () => {
    const sessionCookie = cookie.set('test_session', 'Ab7MNgGyql89hpPalIdgql01gTjkaGb5', {
      expires: new Date('Fri Apr 15 2022 '),
      path: '/',
      secure: true,
      domain: 'https://example.com',
    });

    expect(cookie.set).toBeDefined();
    expect(sessionCookie).toBe(
      `test_session=Ab7MNgGyql89hpPalIdgql01gTjkaGb5; expires=${new Date(
        'Fri Apr 15 2022',
      )}; path=%2F; domain=https%3A%2F%2Fexample.com; secure`,
    );
  });

  it('should throw an error when setting a cookie with a non-string key', () => {
    expect(() => {
      // Explicitly force use of non-string key when setting a cookie to cookie storage
      //@ts-expect-error (2345) : Argument of type 'number' is not assignable to parameter of type 'string'.
      cookie.set(1, 'Hello');
    }).toThrowError(errorsSchema.key);
  });

  it('should throw an error when setting a cookie with a non-string value', () => {
    expect(() => {
      // Explicitly force use of non-string value when setting a cookie to cookie storage
      //@ts-expect-error (2345) : Argument of type 'number' is not assignable to parameter of type 'string'.
      cookie.set('test', 1);
    }).toThrowError(errorsSchema.value);
  });

  it('should throw an error when setting a cookie with a non-object options', () => {
    expect(() => {
      // Explicitly force use of non-object options when setting a cookie to cookie storage
      // @ts-expect-error (2559) : Type '1' has no properties in common with type 'Options'.
      cookie.set('test', 'test', 1);
    }).toThrowError(errorsSchema.options);
  });

  it('should throw an error when setting a cookie with a non-string domain option in the options object', () => {
    expect(() => {
      cookie.set('test', 'test', {
        // Explicitly force use of non-string value for domain option when setting a cookie to cookie storage
        //@ts-expect-error (2345) : Type 'number' is not assignable to type 'string'.
        domain: 1,
      });
    }).toThrowError(errorsSchema.options_domain);
  });

  it('should throw an error when setting a cookie with a non-date expires (value not a date) option in the options object', () => {
    expect(() => {
      cookie.set('test', 'test', {
        // Explicitly force use of value which is not a date on the expires option when setting a cookie to cookie storage
        //@ts-expect-error (2345) : Type 'number' is not assignable to type 'Date'.
        expires: 1,
      });
    }).toThrowError(errorsSchema.options_expires);
  });

  it('should throw an error when setting a cookie with a non-string path option in the options object', () => {
    expect(() => {
      cookie.set('test', 'test', {
        // Explicitly force use of non-string value for path option when setting a cookie to cookie storage
        //@ts-expect-error (2345) : Type 'number' is not assignable to type 'string'.t
        path: 1,
      });
    }).toThrowError(errorsSchema.options_path);
  });

  it('should throw an error when setting a cookie with a non-boolean secure option in the options object', () => {
    expect(() => {
      cookie.set('test', 'test', {
        // Explicitly force use of non-string value for path option when setting a cookie to cookie storage
        //@ts-expect-error (2345) : Type 'number' is not assignable to type 'boolean | undefined'.
        secure: 1,
      });
    }).toThrowError(errorsSchema.options_secure);
  });
});

describe('"cookie.remove"', () => {
  it('should be a properly defined cookie.remove function', () => {
    expect(cookie.remove).toBeDefined();
    expect(typeof cookie.remove).toBe('function');
  });

  it('should return null when a set cookie is deleted from the cookie storage', () => {
    const sessionCookie = cookie.set('test_session', 'Ab7MNgGyql89hpPalIdgql01gTjkaGb5');

    expect(cookie.set).toBeDefined();
    expect(cookie.get).toBeDefined();
    expect(cookie.remove).toBeDefined();
    expect(sessionCookie).toBe('test_session=Ab7MNgGyql89hpPalIdgql01gTjkaGb5');
    expect(cookie.get('test_session')).toBe('Ab7MNgGyql89hpPalIdgql01gTjkaGb5');

    cookie.remove('test_session');
    expect(cookie.get('test_session')).toBeNull();
  });

  it('should throw an error when removing a cookie value with a non-string key', () => {
    expect(() => {
      // Explicitly force use of non-string key when removing a token
      //@ts-expect-error (2345) :argument of type 'number' is not assignable to parameter of type 'string'.
      cookie.remove(1);
    }).toThrowError(errorsSchema.key);
  });
});


src/Cookie.ts
interface Options {
  expires?: Date;
  path?: string;
  domain?: string;
  secure?: boolean;
}

export const encode = global?.window?.encodeURIComponent;
export const decode = global?.window?.decodeURIComponent;

export const errorsSchema = {
  key: '"key" must be a string',
  options: '"options" value must be an object',
  options_domain: '"options.domain" must be a string',
  options_expires: '"options.expires" must be a valid date',
  options_path: '"options.path" must be a string',
  options_secure: '"options.secure" must be a boolean',
  value: '"value" must be a string',
};

export default class Cookie {
  private doc: Partial<Document> | undefined | string;
  constructor(domDocument?: Partial<Document & { cookie: string }> | string) {
    this.doc = domDocument;

    if (typeof domDocument === 'object') {
      this.doc = domDocument;
    }

    if (!this.doc) this.doc = {};
    if (typeof this.doc !== 'object') this.doc = { cookie: '' };
    if (typeof this.doc.cookie !== 'string') this.doc.cookie = '';
  }

  private validateOptions = (options?: Options): void => {
    if (options && typeof options !== 'object') throw new Error(errorsSchema.options);
    if (options && options.domain && typeof options.domain !== 'string') throw new Error(errorsSchema.options_domain);
    if (options && options.expires && !(options.expires instanceof Date)) throw new Error(errorsSchema.options_expires);
    if (options && options.path && typeof options.path !== 'string') throw new Error(errorsSchema.options_path);
    if (options && typeof options.secure !== 'undefined' && typeof options.secure !== 'boolean') {
      throw new Error(errorsSchema.options_secure);
    }
  };

  /**
   * This method will get your cookies from the browser with the specified key
   *
   * @param {string} key - The cookie string used to uniquely identify the cookie in the browser
   *
   * @returns {string | null} - The return value of the found cookie with the specified key provided
   *
   * @example
   * cookie.get("session_value")
   *
   */
  get = (key: string): string | null => {
    if (typeof key !== 'string') throw new Error(errorsSchema.key);

    if (typeof this.doc === 'object' && this.doc.cookie) {
      const splittedCookie = this.doc.cookie.split(/;\cookieString*/);
      for (let cookieIndex = 0; cookieIndex < splittedCookie.length; cookieIndex++) {
        const cookieKeyValues = splittedCookie[0]?.split(';').map((a) => a.trim());

        if (!cookieKeyValues) return null;

        for (let keyIndex = 0; keyIndex <= cookieKeyValues.length; keyIndex++) {
          const cookieKeyValue = cookieKeyValues[keyIndex]?.split('=');
          const cookieKey = decode(cookieKeyValue?.[0] ?? '');

          if (cookieKey === key) return decode(cookieKeyValue?.[1] ?? '');
        }
      }
      return null;
    }

    return null;
  };

  /**
   * This method will set your cookies to the browser the specified key and value provided.
   * If options are provided the value option values will also be set in the cookie storage in the browser
   *
   * @param {string} key - The cookie string used to uniquely identify the cookie in the browser
   *
   * @param {string} value - The value of the cookie you need to set
   *
   * @param {Options} options - The optional parameters to pass when setting the cookie
   *
   * @returns {string | undefined} - The return value of the set cookie
   *
   * @example
   * cookie.set('session_value', 'Uxc70_67gGuHHvAmTy10a', {
   *    expires: new Date(2022, 03, 13),
   *    path: '',
   *    secure: true
   * })
   *
   */
  set = (key: string, value: string, options?: Options): string | undefined => {
    if (typeof key !== 'string') throw new Error(errorsSchema.key);
    if (typeof value !== 'string') throw new Error(errorsSchema.value);
    this.validateOptions(options);

    let opts: Options | undefined = options;

    if (!opts) opts = {};
    let cookieString = encode(key) + '=' + encode(value);
    if (opts.expires) cookieString += '; expires=' + opts.expires;
    if (opts.path) cookieString += '; path=' + encode(opts.path);
    if (opts.domain) cookieString += '; domain=' + encode(opts.domain);
    if (opts.secure) cookieString += '; secure';

    if (typeof this.doc === 'object') {
      this.doc.cookie = cookieString;
    }

    return cookieString;
  };

  /**
   * This method will remove your cookie from the browser with the specified key
   *
   * @param {string} key - The cookie string used to uniquely identify the cookie in the browser
   *
   * @example
   * cookie.remove("session_value")
   */
  remove = (key: string): void => {
    if (typeof key !== 'string') throw new Error(errorsSchema.key);

    let cookieString = encode(key) + '=';
    cookieString += '; expires=' + new Date(0);
    if (typeof this.doc === 'object') {
      this.doc.cookie = cookieString;
    }
  };
}
